pkg/api/mlflow/dao/repositories/namespace_cached.go

```diff
@@ -0,0 +1,186 @@
+			}
+		}
+	}()
+	return &repository, nil
+}
+
+// Create creates new models.Namespace entity.
+func (r NamespaceCachedRepository) Create(ctx context.Context, namespace *models.Namespace) error {
+	if err := r.namespaceRepository.Create(ctx, namespace); err != nil {
+		return eris.Wrap(err, "error creating cached namespace entity")
+	}
+
+	// trigger database event to notify current instance and
+	// other instances to create record in theirs local cache.
+	if err := r.sendEvent(NamespaceEventActionCreated, namespace); err != nil {
+		return eris.Wrap(err, "error sending database event")
+	}
+	return nil
+}
+
+// Update updates existing models.Namespace entity.
+func (r NamespaceCachedRepository) Update(ctx context.Context, namespace *models.Namespace) error {
+	if err := r.namespaceRepository.Update(ctx, namespace); err != nil {
+		return eris.Wrap(err, "error updating cached namespace entity")
+	}
+
+	// trigger database event to notify current instance and
+	// other instances to update record in theirs local cache.
+	if err := r.sendEvent(NamespaceEventActionUpdated, namespace); err != nil {
+		return eris.Wrap(err, "error sending database event")
+	}
+	return nil
+}
+
+// GetByCode returns namespace by its Code.
+func (r NamespaceCachedRepository) GetByCode(ctx context.Context, code string) (*models.Namespace, error) {
+	result, ok := r.cache.Get(code)
+	if ok {
+		return result, nil
+	}
+
+	namespace, err := r.namespaceRepository.GetByCode(ctx, code)
+	if err != nil {
+		return nil, eris.Wrapf(err, "error getting cached namespace by code: %s", code)
+	}
+
+	// trigger database event to notify current instance and
+	// other instances to add record to theirs local cache.
+	if err := r.sendEvent(NamespaceEventActionFetched, namespace); err != nil {
+		return nil, eris.Wrap(err, "error sending database event")
+	}
+	return namespace, nil
+}
+
+// GetByID returns namespace by its ID.
+func (r NamespaceCachedRepository) GetByID(ctx context.Context, id uint) (*models.Namespace, error) {
+	return r.namespaceRepository.GetByID(ctx, id)
+}
+
+// Delete deletes existing models.Namespace entity.
+func (r NamespaceCachedRepository) Delete(ctx context.Context, namespace *models.Namespace) error {
+	if err := r.namespaceRepository.Delete(ctx, namespace); err != nil {
+		return eris.Wrap(err, "error deleting cached namespace entity")
+	}
+
+	// trigger database event to notify current instance and
+	// other instances to remove record from theirs local cache.
+	if err := r.sendEvent(NamespaceEventActionDeleted, namespace); err != nil {
+		return eris.Wrap(err, "error sending database event")
+	}
+	return nil
+}
+
+// List returns all namespaces.
+func (r NamespaceCachedRepository) List(ctx context.Context) ([]models.Namespace, error) {
+	return r.namespaceRepository.List(ctx)
+}
+
+// processEvent process incoming event from database.
+func (r NamespaceCachedRepository) processEvent(data string) error {
```

Comment:
the event subscription logic is pretty complex here -- I think it should get unit test.

Reply:
that's true. unfortunately I don't know how to test it via `unit` tests, because we have to mock this moment:
```
if err := r.db.Exec(
		fmt.Sprintf(`SELECT pg_notify('%s', '%s')`, r.listener.GetChannelName(), data),
	).Error; err != nil {
		return eris.Wrap(err, "error triggering 'pg_notify'")
	}
```
I think it is doable via `integration` tests. added this task #498 as a reminder.